### PR TITLE
Adds `UpdateCollectionPage`

### PR DESCRIPTION
**Pull request by ggirotto**

### CODE REVIEW

File: lib/application/pages/home/collections/update/update_collection_page.dart [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import 'package:memo/application/pages/home/collections/update/update_collection_memos.dart';
+import 'package:memo/application/pages/home/collections/update/update_providers.dart';
+import 'package:memo/application/theme/theme_controller.dart';
+import 'package:memo/application/view-models/home/update_collection_vm.dart';
+import 'package:memo/application/widgets/theme/custom_button.dart';
+import 'package:memo/application/widgets/theme/retry_container.dart';
+import 'package:memo/application/widgets/theme/themed_container.dart';
+import 'package:memo/application/widgets/theme/themed_tab_bar.dart';
+
+enum Segment { details, memos }
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_collection_page.dart:15]: too broad of a name

Comment (matuella) [reply lib/application/pages/home/collections/update/update_collection_page.dart:15]: shouldn't be private?

File: lib/application/pages/home/collections/update/update_collection_page.dart [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+class UpdateCollectionPage extends HookWidget {
+  @override
+  Widget build(BuildContext context) {
+    final vm = readUpdateCollectionVM(context);
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_collection_page.dart:20]: we should watch for the vm itself, as reading in builds is a bad pattern.

File: lib/application/pages/home/collections/update/update_collection_page.dart [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        return strings.memos;
+    }
+  }
+}
+
+class _BottomActionContainer extends HookWidget {
+  const _BottomActionContainer({required this.selectedSegment, required this.onSegmentSwapRequested});
+
+  final Segment selectedSegment;
+  final Function(Segment segment) onSegmentSwapRequested;
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_collection_page.dart:90]:
```suggestion
  final void Function(Segment segment) onSegmentSwapRequested;
```

File: lib/application/pages/home/collections/update/update_collection_page.dart [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      case Segment.details:
+        return strings.details;
+      case Segment.memos:
+        return strings.memos;
+    }
+  }
+}
+
+class _BottomActionContainer extends HookWidget {
+  const _BottomActionContainer({required this.selectedSegment, required this.onSegmentSwapRequested});
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_collection_page.dart:87]: Why nest all of this behavior in these widgets? It makes things much complex to understand. I would simply expose these functionalities to the "consumers" or the parent widgets, so that they can do whatever they want with actions made here in the container, and then, with that, change this bottom container to be simply configurable to all use-cases we need here

File: lib/application/coordinator/routes.dart [hunk 5]
```diff
@@ -87,3 +87,16 @@ class SettingsPath extends AppPath {
   @override
   String get formattedPath => '/$name';
 }
+
+//
+// Update Collection
+//
+class UpdateCollectionPath extends AppPath {
```
Comment (matuella) [inline lib/application/coordinator/routes.dart:94]: Isn't it odd to name it "update" but it's actually creation and update at the same time?

File: lib/application/coordinator/routes.dart [hunk 6]
```diff
@@ -87,3 +87,16 @@ class SettingsPath extends AppPath {
...
+  UpdateCollectionPath({this.collectionId});
+  static const name = 'update_collection';
+
+  final String? collectionId;
```
Comment (matuella) [inline lib/application/coordinator/routes.dart:98]: Don't we already have all the collection metadata once we get to this route? Shouldn't we simply fetch the associated `collection/$id/memos` once we are actually in the route?

File: lib/application/view-models/home/update_collection_vm.dart [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    required this.description,
+    required this.memoMetadata,
+  });
+
+  factory CollectionMetadata.empty() => const CollectionMetadata(name: '', description: '', tags: [], memoMetadata: []);
+
+  final String name;
+  final List<String> tags;
+  final String description;
+  final List<MemoMetadata> memoMetadata;
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:78]: i don't see why we should nest them. Having them composable is better

Comment (matuella) [reply lib/application/view-models/home/update_collection_vm.dart:78]: So that the state can have a Collection and a Memo respective metadata variables

File: lib/application/view-models/home/update_collection_vm.dart [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+      // TODO: Call services to save the collection
+      print('Saving collection...');
+    } on BaseException catch (exception) {
+      state = UpdateCollectionFailedSaving(exception, metadata: loadedState.collectionMetadata);
+    }
+  }
+}
+
+@immutable
+class CollectionMetadata {
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:65]: equatable

File: lib/application/view-models/home/update_collection_vm.dart [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  factory CollectionMetadata.empty() => const CollectionMetadata(name: '', description: '', tags: [], memoMetadata: []);
+
+  final String name;
+  final List<String> tags;
+  final String description;
+  final List<MemoMetadata> memoMetadata;
+}
+
+@immutable
+class MemoMetadata {
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:82]: equatable

File: lib/application/view-models/home/update_collection_vm.dart [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+class UpdateCollectionLoading extends UpdateCollectionState {}
+
+class UpdateCollectionLoaded extends UpdateCollectionState {
+  const UpdateCollectionLoaded({required this.collectionMetadata});
+
+  final CollectionMetadata collectionMetadata;
+
+  bool get hasDetails => true;
+  bool get hasMemos => false;
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:107]: what's going on here?

File: lib/application/view-models/home/update_collection_vm.dart [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  const MemoMetadata({required this.question, required this.answer});
+
+  factory MemoMetadata.empty() => const MemoMetadata(question: '', answer: '');
+
+  final String question;
+  final String answer;
+}
+
+@immutable
+abstract class UpdateCollectionState extends Equatable {
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:92]: We must have an intermediate state for saving, as we will lose all the previous state if not

File: lib/application/widgets/theme/retry_container.dart [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import 'package:memo/application/constants/exception_strings.dart';
+import 'package:memo/application/constants/strings.dart' as strings;
+import 'package:memo/application/widgets/theme/custom_button.dart';
+import 'package:memo/core/faults/exceptions/base_exception.dart';
+
+/// A container that wraps a localized [exception] with a button that, when pressed, calls [onRetry].
+///
+/// This widget is mostly used in scenarios where a page couldn't load its essential data, meaning that there is no
+/// alternative other than retry loading its contents.
+class ExceptionRetryContainer extends StatelessWidget {
```
Comment (matuella) [inline lib/application/widgets/theme/retry_container.dart:12]: Should be a separate PR, but ok.

File: lib/application/widgets/theme/retry_container.dart [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  const ExceptionRetryContainer({required this.onRetry, required this.exception});
+
+  final VoidCallback onRetry;
+  final BaseException exception;
+
+  @override
+  Widget build(BuildContext context) {
+    final textTheme = Theme.of(context).textTheme;
+
+    return Column(
```
Comment (matuella) [inline lib/application/widgets/theme/retry_container.dart:22]: Will this fill the whole screen?

File: lib/application/widgets/theme/retry_container.dart (hunk 12 above)
Comment (matuella) [reply lib/application/widgets/theme/retry_container.dart:12]: Can't we make tests for this widget?

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 1 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_collection_page.dart:15]: > too broad of a name

Do you think? I mean, considering that it's private which means that it applies to this screen only, isn't clear that it references to the tab bar segment?

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 2 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_collection_page.dart:20]: Updated the name to `useUpdateCollectionVM` and added `useProvider` due to the riverpod version

File: lib/application/coordinator/routes.dart (hunk 5 above)
Comment (ggirotto) [reply lib/application/coordinator/routes.dart:94]: That's the pattern we've followed in other projects to name things that were creation and edition at the same time. We could also use `Set` but it seems stranger than `update` to me.

File: lib/application/view-models/home/update_collection_vm.dart (hunk 10 above)
Comment (ggirotto) [reply lib/application/view-models/home/update_collection_vm.dart:107]: Added description and logic

File: lib/application/widgets/theme/retry_container.dart (hunk 13 above)
Comment (ggirotto) [reply lib/application/widgets/theme/retry_container.dart:22]: <details>
<summary>Screenshot</summary>



</details>

File: lib/application/view-models/home/update_collection_vm.dart (hunk 11 above)
Comment (ggirotto) [reply lib/application/view-models/home/update_collection_vm.dart:92]: Ideally this state is `UpdateCollectionLoaded` which may be constantly mutated with the user inputs

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 4 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_collection_page.dart:87]: I've split in different Widgets because the action button may assume different states based on the selected segment and the information that the user added. All use-cases are the following:

- When `Details` segment is selected the button is:
  - Enabled with `Create Collection` title when all the details information is complete and it has at least one memo;
  - Enabled with `Próximo` title when all details information is complete but the user doesn't have at least one memo;
  - Disabled with `Próximo` title when it doesn't have details information complete

- When `Memos` segment is selected the button is:
  - Enabled with `Create Collection` when `details` segment is complete and the user has at least one memo;
  - Disabled with `Create Collection` when there's missing information in `details` or the user doesn't have any memo created. 


Based on all these use-cases, I decided to create distinct buttons for each segment, since it's simpler to understand this division. If we move this use-cases scenarios to a single Widget, we would face an `if hell` scenario with a lot of ternary operations due to the different titles and actions.

File: lib/application/coordinator/routes.dart (hunk 6 above)
Comment (ggirotto) [reply lib/application/coordinator/routes.dart:98]: We have no use-case that moves entities between routes. Do you think that we should introduce `transients` here? I mean, this route would receive an optional `transient` entity named `CollectionMetada`

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 1 above)
Comment (matuella) [reply lib/application/pages/home/collections/update/update_collection_page.dart]: Yeah, it will pollute the global namespace with "Segment"

File: lib/application/widgets/theme/retry_container.dart (hunk 13 above)
Comment (matuella) [reply lib/application/widgets/theme/retry_container.dart]: Shouldn't it be centered and fill all the available height?

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 1 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_collection_page.dart]: But it's private now. It's only visible inside the page

File: lib/application/widgets/theme/retry_container.dart (hunk 13 above)
Comment (ggirotto) [reply lib/application/widgets/theme/retry_container.dart]: I'll update the component parent, because it's centered but the Column is not expanded in the parent

File: lib/application/view-models/home/update_collection_vm.dart (hunk 11 above)
Comment (ggirotto) [reply lib/application/view-models/home/update_collection_vm.dart:100]: I misunderstood your comment at first. Added `UpdateCollectionSaving` and also `UpdateCollectionSaved`

File: lib/application/widgets/theme/exception_retry_container.dart [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+/// This widget is mostly used in scenarios where a page couldn't load its essential data, meaning that there is no
+/// alternative other than retry loading its contents.
+class ExceptionRetryContainer extends StatelessWidget {
+  const ExceptionRetryContainer({required this.onRetry, required this.exception});
+
+  final VoidCallback onRetry;
+  final BaseException exception;
+
+  @override
+  Widget build(BuildContext context) {
```
[the remark this answers is not in the document]
Comment (ggirotto) [reply lib/application/widgets/theme/exception_retry_container.dart:19]: Now it's centered because I've added `Center` wrapper around it in the parent (`UpdateCollectionPage`) widget

File: lib/application/view-models/home/update_collection_vm.dart [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+      state = UpdateCollectionFailedLoading(exception);
+    }
+  }
+
+  @override
+  Future<void> saveCollection() async {
+    final loadedState = state as UpdateCollectionLoaded;
+    try {
+      // TODO: Call services to save the collection
+      print('Saving collection...');
```
Comment (matuella) [inline lib/application/view-models/home/update_collection_vm.dart:61]:
```suggestion
    final loadedState = state as UpdateCollectionLoaded;
    state = UpdateCollectionSaving...
    
    try {
      await Future<void>.delayed(const Duration(seconds: 2));
      // TODO: Call services to save the collection
      state = UpdateCollectionSaved...
```

File: lib/application/pages/home/collections/update/update_providers.dart [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+final updateCollectionId = Provider<String?>((_) => throw UnimplementedError(), name: 'updateCollectionId');
+
+/// Syntax sugar for listening to [UpdateCollectionState] state updates.
+UpdateCollectionState useUpdateCollectionState(WidgetRef ref) {
+  final collectionId = ref.watch(updateCollectionId);
+  return ref.watch(updateCollectionVM(collectionId));
+}
+
+/// Syntax sugar for `ref.watch` the current [UpdateCollectionVM] provider.
+UpdateCollectionVM useUpdateCollectionVM(WidgetRef ref) {
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_providers.dart:15]:
```suggestion
UpdateCollectionState watchUpdateCollectionState(WidgetRef ref) {
  final collectionId = ref.watch(updateCollectionId);
  return ref.watch(updateCollectionVM(collectionId));
}

/// Syntax sugar for `ref.watch` the current [UpdateCollectionVM] provider.
UpdateCollectionVM watchUpdateCollectionVM(WidgetRef ref) {
```

Can't we also make this a non-family provider? wdyt?

File: lib/application/pages/home/collections/update/update_collection_page.dart [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final state = useUpdateCollectionState(ref);
+
+    if (state is! UpdateCollectionLoaded) {
+      return const Center(child: CircularProgressIndicator());
+    }
+
+    final vm = useUpdateCollectionVM(ref);
+    final canSave = state.hasDetails && state.hasMemos;
```
Comment (matuella) [inline lib/application/pages/home/collections/update/update_collection_page.dart:159]: would expose this "canSave" directly from the vm, instead of assuming it from the widget (UI)

File: lib/application/coordinator/routes.dart (hunk 6 above)
Comment (matuella) [reply lib/application/coordinator/routes.dart:98]: That sure is a good question, I'm not sure. This is the same problem that we are facing in multiple other projects, which is: how do we handle "passing data" around multiple pages/states. I don't have a good enough answer now, so I think we could move on with the current solution.

File: lib/application/view-models/home/update_collection_vm.dart [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+    try {
+      // TODO: Call services to load [collectionId] if necessary or create an empty collection metadata
+      await Future<void>.delayed(const Duration(seconds: 2));
+
+      // state = UpdateCollectionLoaded(collectionMetadata: CollectionMetadata.empty());
+      state = UpdateCollectionFailedSaving(
+        UrlException.failedToOpen(),
+        metadata: CollectionMetadata.empty(),
+        memosMetadata: const [],
+      );
```
[the remark this answers is not in the document]
Comment (ggirotto) [reply lib/application/view-models/home/update_collection_vm.dart:50]: Outdated I guess? Also, I would ignore this state management until we have the services good to go, since I'll constantly update it to error states to test the interface

File: lib/application/view-models/home/update_collection_vm.dart (hunk 15 above)
Comment (ggirotto) [reply lib/application/view-models/home/update_collection_vm.dart:61]: Same as https://github.com/olmps/memo/pull/230#discussion_r753225167

File: lib/application/pages/home/collections/update/update_providers.dart (hunk 16 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_providers.dart:15]: I made this changes in a [future PR](https://github.com/olmps/memo/pull/237/files#diff-35ff452a74be7334307fbfa02595615c1d02eec30fc17b6158156376324ce851R1-R10) after the discussion we had in Discord. Please ignore this family providers and functions for now

File: test/utils/fakes.dart [hunk 19]
```diff
@@ -1,6 +1,12 @@
...
 final question = [
   {'question': '?'}
 ];
 final answer = [
   {'answer': '!'}
 ];
+
+class MockCallbackFunction extends Mock {
+  void call();
+}
```
[the remark this answers is not in the document]
Comment (ggirotto) [reply test/utils/fakes.dart:12]: Created a dedicated file

File: lib/application/pages/home/collections/update/update_collection_page.dart (hunk 17 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_collection_page.dart:159]: You mean directly from the state, right? Created a getter there

File: test/utils/fakes.dart (hunk 19 above)
[the remark this answers is not in the document]
Comment (ggirotto) [reply test/utils/fakes.dart:12]: Moved to the proper file

File: lib/application/pages/home/collections/update/update_providers.dart (hunk 16 above)
Comment (ggirotto) [reply lib/application/pages/home/collections/update/update_providers.dart:15]: I removed this auxiliary functions and the `family` providers in a [future PR](https://github.com/olmps/memo/pull/237/files#diff-35ff452a74be7334307fbfa02595615c1d02eec30fc17b6158156376324ce851R1-R10). Please ignore them by now

File: lib/application/coordinator/routes.dart (hunk 6 above)
Comment (ggirotto) [reply lib/application/coordinator/routes.dart:98]: By current solution you mean using just the id or using the transient?

**Review by matuella — APPROVED** [review]
> I would ignore the state management in this function until we have the services good to go, since I'll constantly update it to error states and add unnecessary async Futures to test the interface and I may commit this changes accidentally. What do you think?

That's fine by me, my only concern with those is that we can eventually "introduce" future inconsistencies if not letting those "mocked flows" to a minimal (yet loyal to the expected design). Nonetheless, it's pretty minimal and just like I've said, I'm fine with that.
